src/commands/functions-config-export.ts

```diff
@@ -0,0 +1,383 @@
+ */
+export function convertKey(configKey: string, prefix: string): string {
+  /* prettier-ignore */
+  const baseKey = configKey
+      .toUpperCase()       // 1. Uppercase all characters (e.g. SOME-SERVICE.KEY)
+      .replace(/\./g, "_") // 2. Dots to underscores (e.g. SOME-SERVICE_KEY)
+      .replace(/-/g, "_"); // 3. Dashses to underscores (e.g. SOME_SERVICE_KEY)
+
+  let envKey = baseKey;
+  try {
+    env.validateKey(envKey);
+  } catch (err) {
+    if (err instanceof env.KeyValidationError) {
+      envKey = prefix + envKey;
+      env.validateKey(envKey);
+    }
+  }
+  return envKey;
+}
+
+/**
+ * Convert functions config to environment variables.
+ *   e.g. someservice.key => SOMESERVICE_KEY
+ *
+ * @return {ConfigToEnvResult} Collection of successful and errored conversion.
+ */
+export function configToEnv(configs: Record<string, any>, prefix: string): ConfigToEnvResult {
+  const success = [];
+  const errors = [];
+
+  for (const [configKey, value] of Object.entries(flatten(configs))) {
+    try {
+      const envKey = convertKey(configKey, prefix);
+      success.push({ origKey: configKey, newKey: envKey, value: value as string });
+    } catch (err) {
+      if (err instanceof env.KeyValidationError) {
+        errors.push({
+          origKey: configKey,
+          newKey: err.key,
+          err: err.message,
+          value: value as string,
+        });
+      } else {
+        throw new FirebaseError("Unexpected error while converting config", {
+          exit: 2,
+          original: err,
+        });
+      }
+    }
+  }
+  return { success, errors };
+}
+
+function configsToEnvs(
+  projects: Project[],
+  configs: Record<string, any>[],
+  prefix: string
+): { envs: EnvMap[][]; errMsg: string } {
+  const results = configs.map((config) => configToEnv(config, prefix));
+
+  const errMsg = results
+    .map(({ errors }, idx) => ({ project: projects[idx], errors }))
+    .filter(({ errors }) => errors.length > 0)
+    .map(
+      ({ project, errors }) =>
+        `${project.projectId}:\n` +
+        errors.map((err) => `\t${err.origKey} => ${clc.bold(err.newKey)} (${err.err})`).join("\n") +
+        "\n"
+    )
+    .join("\n");
+  return {
+    envs: results.map((result) => result.success),
+    errMsg,
+  };
+}
+
+async function promptForPrefix(errMsg: string): Promise<string> {
+  logWarning("The following configs keys could not be exported as environment variables:\n");
+  logWarning(errMsg);
+  return await promptOnce(
+    {
+      type: "input",
+      name: "prefix",
+      default: "CONFIG_",
+      message: "Enter a PREFIX to rename invalid environment variable keys:",
+    },
+    { nonInteractive: false }
+  );
+}
+
+function escape(s: string): string {
+  // Escape newlines and tabs
+  let result = s
+    .replace("\n", "\\n")
+    .replace("\r", "\\r")
+    .replace("\t", "\\t")
+    .replace("\v", "\\v");
+  // Escape other escape characters like ' and ".
+  result = result.replace(/(['"])/g, "\\$1");
```

Comment:
Excellent point. I had few iterations of revising this function, and the simplification you suggested evaded me!

Reply:
oh now I remember - it's because `result.replace(/(['"\n\r\t\v])g, "\\$1")` doesn't work!

e.g.

```
> "\n".replace(/([\n])/g, "\\$1")
'\\\n'
> "\n".replace(/([\n])/g, "\$1")
'\n'
> "\n".replace(/\n/g, "\\n")
'\\n'
```

What I want is the last conversion, which is not possible with matched group trick.